Sharps Hazard - Syringes embed when stripped from pockets

pixelpoepleman opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Largely inspired from the hazard of officers getting jabbed and infected when searching junkies, this adds just that. When stripping syringes from peoples pockets, they now embed in the person.

- Syringes now embed into the user when pulled from other's pockets
- This can be countered with Enhanced Retrieval Gloves. 

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

This would potentially set up some more interactions and interesting opportunities for those getting searched, most often by security. This is whether infecting them with some unknowable virus from a used syringe, or one full of chloral hydrate. It gives anyone a funny counter play to a careless pick pocket.

Of course, to also ensure security is not constantly abused by this, enhanced retrieval gloves (regularly available in sec equipment vendor) will protect someone against this. 

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

https://github.com/user-attachments/assets/fb8a4fa2-ec9b-4f5e-af6d-a5b8bc78fc45

</details>

## Changelog
:cl:
add: Added syringes embedding in pickpockets unless protected by thick clothes (seen via examine)

/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

pixelpoepleman committed:
Update strippable.dm

Makes basic syringe embed mechanics to searching pockets filled with syringe.

pixelpoepleman committed:
Merge branch 'master' into PocketSyringe

pixelpoepleman committed:
Full functionality

Moves the given code to only function in pockets. Stops simply grabbing syringes from hands from stabbing you. Now also checks for if the user pick pocketing is a carbon to prevent any weird cases. Further checks for enhanced retrieval gloves to at least give sec an option to not get jabbed. Might be a good idea to make it a type of glove that works instead of just enhanced retrievals.

Rukofamicom commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -182,6 +182,17 @@ GLOBAL_LIST_INIT(strippable_human_layout, list(
+	if(iscarbon(user))
+		var/mob/living/carbon/carbon_user = user
+
+		if(istype(item, /obj/item/reagent_containers/syringe/)  && !istype(carbon_user.gloves, /obj/item/clothing/gloves/tackler))
```
```suggestion
		if(istype(item, /obj/item/reagent_containers/syringe/)  && !istype(carbon_user?.gloves, /obj/item/clothing/gloves/color/black))
```

This should really be any other glove.
These are the gloves security starts with that are thick enough to protect their hands from searing hot lights.

Nakalon replied:
Don't you think that this way this PR is effectively meaningless? As it would only serve to punish tackler glove users (althought i'd love the thought) since secs who don't change their gloves would be protected all the same

Rukofamicom replied:
It's not just security who are affected by this, it's also antagonists stripping medstaff, valid-hunters, etc. 

This would make sec by default protected, but those that upgrade to tackling gloves may find themselves in a bind.

Nakalon replied:
I see but no one starts stripping at the pockets they start stripping the clothing itself. Is this accounting for that also?

Rukofamicom replied:
No it isn't, and that's not a bad point!

Nakalon replied:
I believe it should then! Else this will be bypassed always!

pixelpoepleman replied:
So I will start by saying I fully agree with changing it from enhanced retrievals. The only major reason I chose those was the fact that they are named such which kind of fits with the purpose of protecting someone as they 'retrieve' things from a pocket. 

That said, I am willing to change to whatever type of glove would fit best. If it were to be the typical black gloves or nitriles, then I would happily change to such. My only interesting point on such is that for some strange reason electrical insulating gloves don't protect against heat at all, and from what I understand most _normal_ gloves (other than the black ones) don't either. Meanwhile, something like the wardens krav gloves do interestingly protect from heat, even though they are fingerless. This does bring up a good point though, this change should most directly affect wardens. If anything, it should be a different type of glove than the krav gloves (maybe just punish fingerless gloves?)

To further the point about full stripping uniforms, this is something I had considered and elected to not deal with for the moment. This is because it is actually noted in space law that security should not full strip and give someone prisoner oranges unless they are due for a permanent sentence! If it would still be best to make this change, I will try my best to make it so.

Nakalon replied:
> To further the point about full stripping uniforms, this is something I had considered and elected to not deal with for the moment. This is because it is actually noted in space law that security should not full strip and give someone prisoner oranges unless they are due for a permanent sentence! If it would still be best to make this change, I will try my best to make it so.

I think you have a good point here, however players will powergame the shit out of this kind of stuff.

Say you make the retrieval gloves protect from this. Then ALL secoffs will use those gloves, because for the price of NOTHING they are now protected from this.

Also, since ruko said this would be good because it also affects valid hunters and antags then you must account for stripping because thats what they will do hahaha, good luck friend!

Dejaku51 commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -182,6 +182,17 @@ GLOBAL_LIST_INIT(strippable_human_layout, list(
+	if(iscarbon(user))
+		var/mob/living/carbon/carbon_user = user
+
+		if(istype(item, /obj/item/reagent_containers/syringe/)  && !istype(carbon_user.gloves, /obj/item/clothing/gloves/tackler))
```
maybe instead of using istype and specifying exact gloves use some flags, or make it based on armor values or smth

EDIT: current code for protecting from syringes is:
A) `get_armor().get_rating(MELEE) >= 10 || get_armor().get_rating(BULLET) >= 10` - protect from normal syringes
B) `clothing_flags & THICKMATERIAL` - protect from piercing syringes

PowerfulBacon replied:
Use the clothing flag thick material instead of doing an istype check

pixelpoepleman committed:
Merge branch 'master' into PocketSyringe

pixelpoepleman committed:
Merge branch 'BeeStation:master' into PocketSyringe

pixelpoepleman committed:
PR Review Compliance

Now compliant with all the PR reviews. Changes the types of gloves that are resistant to the syringes, and makes a few of such. Also adds functionality to fullstripping jumpsuits. Piercing syringes also go through the specific gloves when fullstripping.

PowerfulBacon commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -77,6 +77,54 @@ GLOBAL_LIST_INIT(strippable_human_layout, list(
+
+	var/result = start_unequip_mob(item, source, user, strip_delay = POCKET_STRIP_DELAY, hidden = TRUE)
+
+	return result
```
A lot of this is copy paste from the parent proc. Call ..() and if that returns TRUE, then do the extra behaviour
```suggestion
/datum/strippable_item/mob_item_slot/jumpsuit/start_unequip(atom/source, mob/user)
	var/obj/item/item = get_item(source)

	if (!..())
		return FALSE

	if(iscarbon(user))
		var/mob/living/carbon/carbon_user = user
		var/mob/living/carbon/source_pocket = source
		var/obj/item/pocket_item = source_pocket.get_item_by_slot(ITEM_SLOT_LPOCKET)
		var/obj/item/reagent_containers/syringe/syringeitem = null

		if(istype(pocket_item, /obj/item/reagent_containers/syringe))
			syringeitem = pocket_item

		else
			pocket_item = source_pocket.get_item_by_slot(ITEM_SLOT_RPOCKET)
			if (istype(pocket_item, /obj/item/reagent_containers/syringe))
				syringeitem = pocket_item

		if(syringeitem)

			if((!carbon_user.gloves || !(carbon_user.gloves.clothing_flags & THICKMATERIAL)) || (syringeitem.type == /obj/item/reagent_containers/syringe/piercing))
				finish_unequip_mob(syringeitem, source, user)
				syringeitem.embed(user)
				user.visible_message(span_danger("You see [user] try to rip off [source]'s jumpsuit and scream in pain!"), span_userdanger("A syringe embeds itself in your hand!"))
				user.emote("scream")
				return FALSE

	var/result = start_unequip_mob(item, source, user, strip_delay = POCKET_STRIP_DELAY, hidden = TRUE)

	return result
```
A lot of this is copy paste from the parent proc. Call ..() and if that returns TRUE, then do the extra behaviour

PowerfulBacon commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -77,6 +77,54 @@ GLOBAL_LIST_INIT(strippable_human_layout, list(
+				syringeitem.embed(user)
+				user.visible_message(span_danger("You see [user] try to rip off [source]'s jumpsuit and scream in pain!"), span_userdanger("A syringe embeds itself in your hand!"))
+				user.emote("scream")
+				return FALSE
```
This behaviour should be generic, add a new proc on /obj/item/proc/on_start_unequip (or something similar) and move this syringe behaviour to be entirely contained in the syringe type. Then anything can add additional behaviour similar to this. Generally comparing types is seen as bad code as it isn't modular and can't be expanded upon in the future.

pixelpoepleman committed:
Bacon Compliance

Adds new proc that gets called in start_unequip. Used for checking items for effects when being stripped. Moved relevant effects of embedding to the syringe specific version of this proc. Uniforms now also run this check for items in the pockets when being stripped.

PowerfulBacon commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -312,3 +312,28 @@
 	fill_icon_state = "syringe_crude"
 	fill_icon_thresholds = list(5, 10, 15)
+
+/obj/item/reagent_containers/syringe/on_start_stripping(mob/source, mob/user, item_slot, obj/item)
```
You don't need item here, since this thing itself is the item (src).

PowerfulBacon replied:
```suggestion
/obj/item/reagent_containers/syringe/on_start_stripping(mob/source, mob/user, item_slot)
```

PowerfulBacon commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -312,3 +312,28 @@
+		return FALSE
+
+	var/mob/living/carbon/carbon_user = user
+	var/obj/item/reagent_containers/syringe/syringeitem = item
```
```suggestion
```

PowerfulBacon commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -312,3 +312,28 @@
+
+	if(item_slot == ITEM_SLOT_LPOCKET || item_slot == ITEM_SLOT_RPOCKET)
+		if(!carbon_user.gloves || !(carbon_user.gloves.clothing_flags & THICKMATERIAL))
+			finish_unequip_mob(item, source, user)
```
```suggestion
			finish_unequip_mob(src, source, user)
```

PowerfulBacon commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -312,3 +312,28 @@
+	if(item_slot == ITEM_SLOT_LPOCKET || item_slot == ITEM_SLOT_RPOCKET)
+		if(!carbon_user.gloves || !(carbon_user.gloves.clothing_flags & THICKMATERIAL))
+			finish_unequip_mob(item, source, user)
+			syringeitem.embed(user)
```
```suggestion
			embed(user)
```

PowerfulBacon commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -312,3 +312,28 @@
+			return TRUE
+	else
+		if(item_slot == ITEM_SLOT_ICLOTHING)
+			if((!carbon_user.gloves || !(carbon_user.gloves.clothing_flags & THICKMATERIAL)) || (syringeitem.type == /obj/item/reagent_containers/syringe/piercing))
```
```suggestion
			if((!carbon_user.gloves || !(carbon_user.gloves.clothing_flags & THICKMATERIAL)) || proj_piercing)
```
You can use the proj_piercing variable on syringe to check if it is piercing or not.

PowerfulBacon commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -312,3 +312,28 @@
+	else
+		if(item_slot == ITEM_SLOT_ICLOTHING)
+			if((!carbon_user.gloves || !(carbon_user.gloves.clothing_flags & THICKMATERIAL)) || (syringeitem.type == /obj/item/reagent_containers/syringe/piercing))
+				finish_unequip_mob(syringeitem, source, user)
```
```suggestion
				finish_unequip_mob(src, source, user)
```

PowerfulBacon commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -312,3 +312,28 @@
+		if(item_slot == ITEM_SLOT_ICLOTHING)
+			if((!carbon_user.gloves || !(carbon_user.gloves.clothing_flags & THICKMATERIAL)) || (syringeitem.type == /obj/item/reagent_containers/syringe/piercing))
+				finish_unequip_mob(syringeitem, source, user)
+				syringeitem.embed(user)
```
```suggestion
				embed(user)
```

PowerfulBacon commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -182,6 +184,9 @@ GLOBAL_LIST_INIT(strippable_human_layout, list(
 	user.log_message(log_message, LOG_ATTACK, color="red", log_globally=FALSE)
 	item.add_fingerprint(src)
 
+	if(item.on_start_stripping(source, user, source_pocket.get_slot_by_item(item), item))
```
```suggestion
	if(item.on_start_stripping(source, user, source_pocket.get_slot_by_item(item)))
```

PowerfulBacon commented on code/modules/clothing/under/_under.dm:
```diff
@@ -387,3 +387,21 @@
+		if(pocket_item)
+			result = pocket_item.on_start_stripping(source, user, ITEM_SLOT_ICLOTHING, pocket_item)
+
+	return result
```
```suggestion
/obj/item/clothing/under/on_start_stripping(mob/source, mob/user, item_slot)
	if(!iscarbon(user))
		return FALSE

	var/mob/living/carbon/source_pocket = source
	var/obj/item/pocket_item = source_pocket.get_item_by_slot(ITEM_SLOT_LPOCKET)

	if(pocket_item && pocket_item.on_start_stripping(source, user, ITEM_SLOT_ICLOTHING))
		return TRUE

	pocket_item = source_pocket.get_item_by_slot(ITEM_SLOT_RPOCKET)
	if(pocket_item && pocket_item.on_start_stripping(source, user, ITEM_SLOT_ICLOTHING))
		return TRUE

	return FALSE
```

PowerfulBacon commented on code/game/objects/items.dm:
```diff
@@ -1624,3 +1624,7 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 /obj/item/proc/perform_strip_actions(action_key, mob/actor)
+
+// For item specific checks on strip start. Return true to interrupt stripping, return false to continue stripping.
+/obj/item/proc/on_start_stripping(mob/source, mob/user, item_slot, obj/item)
```
```suggestion
/obj/item/proc/on_start_stripping(mob/source, mob/user, item_slot)
```

PowerfulBacon commented on code/datums/elements/strippable.dm:
```diff
@@ -181,6 +183,9 @@
 	user.log_message("[key_name(source)] is being stripped of [item.name] by [key_name(user)]", LOG_ATTACK, color="red", log_globally=FALSE)
 	item.add_fingerprint(src)
 
+	if(item.on_start_stripping(source, user, source_pocket.get_slot_by_item(item), item))
```
```suggestion
	if(item.on_start_stripping(source, user, source_pocket.get_slot_by_item(item)))
```

pixelpoepleman committed:
Formatting Fixes

Makes things cleaner and look nicer with better functionality.
